### DISCUSSION TITLE
AttributeError: 'async_generator' object has no attribute 'add' when using Dependency Injection

### DISCUSSION THREAD

**Discussion by shoang22** [fastapi/fastapi#11938] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# app/dependencies.py

from typing import Annotated, AsyncGenerator 

from sqlmodel.ext.asyncio.session import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from fastapi import Depends

from app.settings import settings


async def get_db() -> AsyncGenerator[AsyncSession, None]:
    db_config = ConnectionConfig(
        service="mysql",
        driver="asyncmy",
        user=settings.MYSQL_USER,
        password=settings.MYSQL_PASSWORD,
        host=settings.MYSQL_HOST,
        database=settings.MYSQL_DATABASE,
    )
    engine = create_async_engine(
            get_db_url(config=db_config), echo=settings.DEVELOPMENT
        )
    async with AsyncSession(engine) as session:
        yield session


# test_db.py

from typing import AsyncGenerator, Iterator
from contextlib import asynccontextmanager

import pytest
from fastapi import status
from fastapi.testclient import TestClient
from sqlmodel import SQLModel
from sqlmodel.ext.asyncio.session import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine

from app.dependencies import get_db
    

@pytest.fixture
async def session() -> AsyncGenerator[AsyncSession, None]:
    engine = create_async_engine("sqlite+aiosqlite://")
    async with engine.connect() as cxn, cxn.begin() as txn:
        await cxn.run_sync(SQLModel.metadata.create_all)
        async_session = AsyncSession(
            bind=cxn, join_transaction_mode="create_savepoint"
        )
        yield async_session

        await txn.rollback()
    await engine.dispose()


@pytest.fixture  
def client(session: AsyncSession):  
    def get_session_override():  
        return session

    app = create_app()
    app.dependency_overrides[get_db] = get_session_override  

    with TestClient(app) as c:
        yield c  

    app.dependency_overrides.clear()  


@pytest.mark.anyio
def test_create_attestation(
    attestations: Iterator[Attestation], client: TestClient
):  
    for attestation in attestations:
        payload = {"topic": "some topic"}
        resp = client.post(url="/v1/upload", json=payload)
        assert resp.status_code == status.HTTP_200_OK


# app/routers.py

@router.post(
    path="/upload",
)
async def upload(
    req: UploadRequest, db: AsyncSession = Depends(get_db)
):
    row = Row(**req.model_dump())
    db.add(row)
    await db.commit()
```

### Description

I'm trying to run a test suite on FastAPI endpoints that use a database connection. I'm using FastAPI's dependency injection on the endpoint, which I attempt to override with a test database during testing.

When running the endpoints directly, the dependency injection works, but in the test suite, I get the following error:

```AttributeError: 'async_generator' object has no attribute 'add'```

I've tried to modify the fixture scopes, as well as trying to iterate through the async_generator using `__anext__`, but I'm still getting the same error. Any help would be appreciated.

### Operating System

macOS

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

3.12.2

**Comment by estebanx64** [comment, marked as answer]
Hi!

It seems that your test configuration setup isn't right.

I would recommend you to read this section https://fastapi.tiangolo.com/advanced/async-tests/
There is even another discussion about how to set async test here as an example https://github.com/fastapi/fastapi/discussions/10958

**Reply by shoang22** [reply]
Thank you! FYI using `pytest_asyncio.fixture` also did the trick.
